### PR TITLE
Adding frequency to SubscribeAttributes

### PR DESCRIPTION
**Pull request by stevenhartley**
Frequency will be used to throttle published events that are to be sent between devices. This is commonly used when cloud/mobile applications wish to subscribe to high frequency vehicle topics

#80

### CODE REVIEW

**Comment by sfactor33** [comment]
I agree with using the usub attribute of max_frequency for publishing topic to the remote.

**Review by jefftindall — APPROVED** [review]

**Review by sfactor33 — APPROVED** [review]
Approved.

**Review by frankghe — CHANGES_REQUESTED** [review]
why double, and not unsigned it?

**Comment by tamarafischer** [comment]
What you are trying to achieve is backpressure and I am pretty sure that adding a throttle attribute is not the best way to implement this architecture pattern.

**Comment by stevenhartley** [comment]
> What you are trying to achieve is backpressure and I am pretty sure that adding a throttle attribute is not the best way to implement this architecture pattern. If you scale up and add more resources, does this mean I have to change the subscription?

@tamarafischer what do you propose instead?

**Comment by stevenhartley** [comment]
@tamarafischer , here is some valueable info I got from ChatGPT:

_Throttling and backpressure are two different approaches used in distributed messaging systems to handle situations where the rate of message production exceeds the rate of message consumption._

- _Throttling: Throttling is a mechanism used to control the rate at which messages are produced or sent. It involves limiting the rate of message production to prevent overwhelming the system or the downstream consumers. Throttling can be implemented by introducing delays between message production or by enforcing quotas on the number of messages that can be produced within a certain time period. Throttling helps to regulate the flow of messages and prevent resource exhaustion._

- _Backpressure: Backpressure is a mechanism used to control the rate of message consumption or processing. It involves signaling the message producer to slow down or stop producing messages when the consumer or downstream system is unable to keep up with the incoming messages. Backpressure can be implemented by using flow control mechanisms such as buffer sizes, acknowledgments, or explicit signals from the consumer to the producer. Backpressure helps to balance the load between the producer and consumer, ensuring that the system operates within its capacity._

_In summary, throttling focuses on controlling the rate of message production, while backpressure focuses on controlling the rate of message consumption. Throttling is typically implemented by the producer, while backpressure is typically implemented by the consumer. Both mechanisms are used to prevent overload and maintain stability in distributed messaging systems._

Our intention is to control the rate of production of events from device-2-device and not control the rate of consumption so the above solution should apply. now the question remains should we define max_frequency in hertz (double) or in miliseconds (uint32), given miliseconds could take up less space to send, it might be more intuative to make this uint32 ms. @jefftindall @sfactor33 what say you?

**Comment by sfactor33** [comment]
Throttling is the desired behavior here.  Backpressure control is not the ask per the design discussions.  
The desired effect is to be able to take a pub message with a high publishing rate (say 100ms) and have a subscription based process for remotes to reducing the message rate (say to 5s rate).  The remote subscriber would have determined that this meets their needs and accepts missing the pub events.  There are still other subscribers locally that need the pub messages at the "full" rate, which is why we wouldn't want to have the publisher reduce its pub rate.  A tool in the toolbox that meets some needs.  The max_frequency attribute would be selected at design time where appropriate.

**Comment by jefftindall** [comment]
TL;DR I vote for hertz.

> now the question remains should we define max_frequency in hertz (double) or in miliseconds (uint32), given miliseconds could take up less space to send, it might be more intuative to make this uint32 ms.

Personally, I think the units should match the name.

`max_frequency` should be hertz because hertz is the SI unit for frequency.
`sampling_period` could be milliseconds

I think `max_frequency` is intuitive as an attribute in this context and therefore would favor hertz.  Given the frequency (no pun intended) of how often we expect subscription requests, I don't think optimizing for space (and extra byte) is a big deal.  If this were the data payloads for high frequency data, I would probably have a different opinion.

**Comment by frankghe** [comment]
I agree with the idea that frequency should be Hz, and sampling period should be ms. We have 2 main use-cases:
- cloud to vehicle subscription: in this case, sampling period is surely above 1s, and most likely granularity is seconds, so either frequency or sampling period equally work well
- in vehicle subscription (likely across ECUs): in this case, we will often have sub-second sampling periods and using Hz is very unintuitive

I therefore vote for sampling period, using uint or long.

File: uprotocol/core/usubscription/v3/usubscription.proto [hunk 1]
```diff
@@ -118,6 +118,16 @@ message SubscribeAttributes {
...
+
+  // Maximum Frequency (measured in Hertz) the subscriber wishes to receive events 
+  // for remote only topics. Device dispatchers (i.e. streamers) use this
+  // attribute to reduce the publication rates of events sent between devices.
+  // This attribute is commonly used for mobile/cloud subscribing to vehicle topics
+  // that are published at a high frequency.<br>
+  // If the frequency set by the subscriber is higher than the original publishers
+  // publication frequency, the attribute is ignored.<br>
+  // **__NOTE:__** If this attribute is missing, the frequency is set by the publisher.
+  optional double max_frequency = 3;
```
Comment (PLeVasseur) [inline uprotocol/core/usubscription/v3/usubscription.proto:130]: Based on the feedback, seems like this could morph to:
```protobuf
  // Desired Sampling Period (measured in milliseconds) the subscriber wishes to receive events 
  // for remote only topics. Device dispatchers (i.e. streamers) use this
  // attribute to reduce the publication rates of events sent between devices.
  // This attribute is commonly used for mobile/cloud subscribing to vehicle topics
  // that are published at a high rate.<br>
  // If the desired sampling period set by the subscriber is lower than the original publisher's
  // publication period, the attribute is ignored.<br>
  // **__NOTE:__** If this attribute is missing, the sampling period is set by the publisher.
  optional uint32 desired_sampling_period = 3;
```

Comment (frankghe) [reply uprotocol/core/usubscription/v3/usubscription.proto:130]: good for me!

Comment (stevenhartley) [reply uprotocol/core/usubscription/v3/usubscription.proto:130]: Thanks peter, I'll change a little but then merge, I'll need +1 from a committer to merge.

Comment (PLeVasseur) [reply uprotocol/core/usubscription/v3/usubscription.proto]: Just a point of curiosity for me -- why would we choose an int32 over a uint32 if we know that the value will always be unsigned?

**Review by frankghe — APPROVED** [review]
